### DISCUSSION TITLE
Create a smooth curve through known points

### DISCUSSION THREAD

**Discussion by raydeleu** [sgenoud/replicad#42] [How To]
Today I tried to model a hook for Apple Earpods. These hooks, published on Thingiverse by Tom Lombardi (see https://www.thingiverse.com/thing:153374), are very practical to keep them fixed to your ears while running. 

While modelling the curve needed to create these hooks, I tried several methods. Perhaps it is helpful for others as well to understand the difference between the methods to create curves. 

I started with drawing the rough shape of the curve. 

![hook-sketch](https://user-images.githubusercontent.com/38007983/177052401-2d5e1e08-5d81-413a-a18d-174d29623a15.jpeg)

From this sketch I derived some "throughPoints" for the curve. The following code shows the result of some experiments to create a curve through these points. I found the following: 

* Using `smoothSplineTo` requires the least amount of commands but results in a very strange curve. Obviously the tangents at the beginning and the end of the curve are not related to any other point in the curve but are set to 0 degrees. Therefore it is almost obligatory to add a configuration statement such as `{startTangent:180,endTangent:125}`. Still the resulting curve is difficult to control and does not look very smooth; 

* The `bezierCurve` results in a nice smooth curve overall, but as all points except for the last point are used only as control points, a lot of trial and error is needed to adjust these points to find the correct curve. 

![hook-comparison](https://user-images.githubusercontent.com/38007983/177053316-6ec5f0f0-8880-4394-bea5-4bb6844dbe74.jpg)

But the control points are really very weird to arrive at this shape: 

<img width="733" alt="bezierCurve2" src="https://user-images.githubusercontent.com/38007983/177054131-6cb0058f-f5a6-4899-884c-b0594deb8055.png">

* With a `quadraticBezierCurve` it is relatively easy to force the curve through the `throughPoints`. Only one control point has to be added per point to define the tangents. The downside is that it is not possible to influence both the tangent and the path of the curve. 

* For this application, the `cubicBezierCurve` is the best option. The curve can be forced through the `throughPoints` and with the two control points per segment it is easy to adjust both the tangent and the shape of the curve. 

<img width="698" alt="control-points" src="https://user-images.githubusercontent.com/38007983/177053356-bdb99d6f-ed33-4e38-9a1f-8e1e9588ef90.png">

The code below contains all experiments. Using the Visualiser you can compare and adjust the shape of all curves. 

```` javascript
const defaultParams = {                // setting the value of the parameters
    earbudStemWidth:        4.5,
    earbudStemThickness:    4.0,
    earbudStemHeight:       20.0
    }
  
  // next lines allow intellisense help in VS Code
  /** @typedef { typeof import("replicad") } replicadLib */
  /** @type {function(replicadLib, typeof defaultParams): any} */
  
  function main(
   { draw, drawRoundedRectangle},   // functions used within the main program
   { earbudStemWidth,earbudStemThickness,earbudStemHeight} )  // parameters to adjust the model
  {
    // actual points of the contour
    let p0 = [0,0]
    let p1 = [20,10]
    let p2 = [28,25]
    let p3 = [10,42]
    let p4 = [-12,32]
    let p5 = [-20,35]

    // increased number of points for Bezier
    // still tweak necessary as line only approaches points
    let t1 = [15,3]
    let t2 = [25,10]
    let t3 = [30,15]
    let t4 = [32,25]
    let t5 = [33,30]
    let t6 = [31,40]
    let t7 = [26,45]
    let t8 = [15,47]
    let t9 = [0,45]
    let t10 = [-5,38]
    let t11 = [-10,35]
    let t12 = [-12,32]
    let t13 = [-15,33]
    let t14 = [-20,35]

    // points tweaked to get right shape
    let b1 = [30,5]
    let b2 = [55,35]
    let b3 = [3,80]
    let b4 = [-2,15]
    let b5 = [-20,35]

    // control points for cubicBezier
    //let p0 = [0,0]
    // let p1 = [20,10]
    // let p2 = [28,25]
    let c2 = [28,0]
    // let p3 = [10,42]
    let c3 = [28,42]
    // let p4 = [-12,32]
    let c4 = [0,42]
    // let p5 = [-20,35]
    let c5 = [-8,42]
    let c6 = [-8,25]

    // points cubicBezierCurve
    // throughPoints are tp1 .. tp2
    // controlPoints are cp1s, cp1e (control point start/end)

    let tp1 = [28,25]
    let cp1s = [15,0]
    let cp1e = [28,15]

    let tp2 = [10,42]
    let cp2s = [28,35]
    let cp2e = [20,42]

    let tp3 = [-12,32]
    let cp3s= [0,42]
    let cp3e= [-8,32]

    let tp4 = [-20,35]
    let cp4s= [-14,32]
    let cp4e= [-16,32]

    // curve smoothSpline, no tangents specified
    let hookCurve1 = draw().smoothSplineTo(p1).smoothSplineTo(p2).smoothSplineTo(p3).smoothSplineTo(p4).smoothSplineTo(p5).done().sketchOnPlane("XY")

    // curve smoothSpline, tangents specified
    // difficult to control "sag" between points
    let hookCurve2 = draw().smoothSplineTo(p1,{startTangent:0,endTangent:45}).smoothSplineTo(p2,{startTangent:45,endTangent:90}).smoothSplineTo(p3,{startTangent:90,endTangent:180}).smoothSplineTo(p4,{startTangent:180,endTangent:180}).smoothSplineTo(p5,{startTangent:180,endTangent:125}).done().sketchOnPlane("XY")

    // curve Beziercurve using throughpoints as controlpoints
    let hookCurve3 = draw().bezierCurveTo(p5,[p1,p2,p3,p4]).done().sketchOnPlane("XY")

    // bezierCurve with large number of points
    // still not sufficient to follow through points
    let hookCurve4 = draw().bezierCurveTo(t14,[t1,t2,t3,t4,t5,t6,t7,t8,t9,t10,t11,t12,t13]).done().sketchOnPlane("XY")

    // curve Beziercurve where points are massively tweaked
    // trial and error necessary
    let hookCurve5 = draw().bezierCurveTo(b5,[b1,b2,b3,b4]).done().sketchOnPlane("XY")

    // curve based largely on quadraticBezierCurve
    // difficult to control curve with only one control point
    let hookCurve6 = draw().quadraticBezierCurveTo(p2,c2).quadraticBezierCurveTo(p3,c3).cubicBezierCurveTo(p5,c5,c6).done().sketchOnPlane("XY")

    let hookCurve7 = draw().cubicBezierCurveTo(tp1,cp1s,cp1e).cubicBezierCurveTo(tp2,cp2s,cp2e).cubicBezierCurveTo(tp3,cp3s,cp3e).cubicBezierCurveTo(tp4,cp4s,cp4e).done().sketchOnPlane("XY")

    let shapeArray = [
        {shape: hookCurve1, name:"smoothSplineNoTangent",color: "black"}
        ,{shape: hookCurve2, name:"smoothSplineTangent",color: "red"}
        ,{shape: hookCurve3, name:"BezierCurveStd", color: "orange"}
        ,{shape: hookCurve4, name:"BezierAdaptedMorePoints", color: "steelblue"}
        ,{shape: hookCurve5, name:"BezierCurveAdapted", color: "purple"}
        ,{shape: hookCurve6, name:"quadraticBezier", color: "blue"}
        ,{shape: hookCurve7, name:"cubicBezier", color: "fuchsia"}
    ]   

    return  shapeArray
   }
````

**Comment by lf94** [comment]
You will always have issues this problem when using any bezier curve because they aren't designed this way.

Instead https://en.wikipedia.org/wiki/Centripetal_Catmull%E2%80%93Rom_spline is needed and is much more complex but 100% solves the problem.

I'm looking around at OpenCascade to see if it has one. https://dev.opencascade.org/doc/refman/html/search.php?query=spline+-bspline is the query I used. 

"Wilson-Fowler" (Spline type 4) seems to be functional equivalent to Catmull rom...

https://dev.opencascade.org/doc/refman/html/class_i_g_e_s_geom___spline_curve.html

**Reply by raydeleu** [reply]
In CascadeStudio there is a `BSpline` function that I compared earlier to the bezier function. In the image below, the black curve is created using the `.BSplineTo` method, whereas the white curve is created using the `.BezierTo` method. The little cylinders show the position of the control points used to generate both curves.  

![spline20](https://user-images.githubusercontent.com/38007983/177056769-63285300-719c-4a4e-804a-cdc8bc1e8202.png)

The code for the `BSplineTo` method is: 

````javascript
/* Constructs a BSpline from the previous point through this set of points */
  this.BSplineTo = function (bsplinePoints) {
    this.argsString += ComputeHash(arguments, true);
    let ptList = new oc.TColgp_Array1OfPnt(1, bsplinePoints.length+1);
    ptList.SetValue(1, this.lastPoint);
    for (let bInd = 0; bInd < bsplinePoints.length; bInd++){
      let ctrlPoint = convertToPnt(bsplinePoints[bInd]);
      ptList.SetValue(bInd + 2, ctrlPoint);
      this.lastPoint = ctrlPoint;
    }
    let handle         = new oc.GeomAPI_PointsToBSpline(ptList  ).Curve();
    let lineEdge       = new oc.BRepBuilderAPI_MakeEdge(handle  ).Edge() ;
    this.wireBuilder.Add(new oc.BRepBuilderAPI_MakeWire(lineEdge).Wire());
    this.currentIndex++;
    return this;
  }
````
It seems that it uses the OpenCascade function `GeomAPI_PointsToBSpline(ptList)`. 

I am very interested in this function, as another experiment I have in mind is to create an airfoil shape. I intended to use the airfoil coordinates database available at https://m-selig.ae.illinois.edu/ads.html. A completely different approach could be to use an SVG image of the airfoil (see https://m-selig.ae.illinois.edu/ads/frink/airfoils.svg) and see whether your conversion tools are able to convert from this SVG to a shape in Replicad. I wonder which method your code would use to convert these shapes.

**Reply by lf94** [reply]
Yes, my tool is able to convert it. It's different from the original problem, because the original problem is "make a curve that passes through these points" and not "make a make curve that matches these curves". BSpline gets you very close for sure! But I think Wilson-Fowler or Catmull-Rom get you even closer _from what I understand_. It appears BSpline is "good enough" solution though!

**Comment by sgenoud** [comment]
Note that I have made the `smoothSpline` so that it works like a visual bézier curve editor. You only need to specify the tangent of the first point (if non 0), then it automatically uses the tangent of the previous segment (it does not need to be a spline either). Additionally you can play with the `endFactor` and startFactor` to give more or less weight to the start or the end of the bézier curve.

All this said, I will provide a `drawPointsInterpolation` (which is what `drawParametricFunction` uses anyway. Stay tuned!

**Reply by sgenoud** [reply]
The new version of the visualiser (based on 0.12.2) supports `drawPointsInterpolation`

```js
const main = ({ drawPointsInterpolation } ) => {
  const c = [
    [0, 0],
    [20, 10],
    [28, 25],
    [10, 42],
    [-12, 32],
    [-20, 35],
  ];

  return drawPointsInterpolation(c).sketchOnPlane();
};
```

**Reply by raydeleu** [reply]
With the smoothSpline I found the following result: 

<img width="479" alt="smoothspline" src="https://user-images.githubusercontent.com/38007983/177213673-3f2111d4-1b30-4073-b01d-40ad0500445c.png">

```` javascript
let hookCurve1 =draw()
.smoothSplineTo(p1)
.smoothSplineTo(p2)
.smoothSplineTo(p3)
.smoothSplineTo(p4)
.smoothSplineTo(p5)
.done()
.sketchOnPlane("XY")
````

If I combine all points together in an array it seems to disregard the array of intermediate points. As reported earlier, with the tangents I get an acceptable result but it is difficult to control the roundness of the curve. The drawPointsInterpolation indeed results in an acceptable shape. I guess to control this curve any further I would have to use more intermediate points. 

<img width="551" alt="comparison-smoothspline" src="https://user-images.githubusercontent.com/38007983/177215939-40750b0b-2e2b-4408-9964-09e0f3d82cc7.png">

**Reply by sgenoud** [reply]
I have played a little bit with the example to show you how the start (and end) factor can make it easy to control the shape of your curve.

```js
const defaultParams = {
  p1Start: 1.5,
  p1End: 1,
  p2Start: 1,
  p2End: 2,
  p3Start: 2,
  p3End: 1,
  p4Start: 2,
  p4End: 1,
  p5Start: 1,
  p5End: 1,
};

const main = (
  { draw, drawPointsInterpolation },
  {
    p1Start,
    p1End,
    p2Start,
    p2End,
    p3Start,
    p3End,
    p4Start,
    p4End,
    p5Start,
    p5End,
  }
) => {
  let p0 = [0, 0];
  let p1 = [20, 10];
  let p2 = [28, 25];
  let p3 = [10, 42];
  let p4 = [-12, 32];
  let p5 = [-20, 35];

  const c1 = draw()
    .smoothSplineTo(p1, {
      startTangent: 0,
      endTangent: 45,
      startFactor: p1Start,
      endFactor: p1End,
    })
    .smoothSplineTo(p2, {
      endTangent: 90,
      startFactor: p2Start,
      endFactor: p2End,
    })
    .smoothSplineTo(p3, {
      endTangent: 180,
      startFactor: p3Start,
      endFactor: p3End,
    })
    .smoothSplineTo(p4, {
      endTangent: 180,

      startFactor: p4Start,
      endFactor: p4End,
    })
    .smoothSplineTo(p5, {
      endTangent: 125,
      startFactor: p5Start,
      endFactor: p5End,
    })
    .done()
    .sketchOnPlane();

  const c2 = drawPointsInterpolation([p0, p1, p2, p3, p4, p5]).sketchOnPlane();
  return [c1, c2];
};
```

<img width="684" alt="Capture d’écran 2022-07-05 à 08 44 51" src="https://user-images.githubusercontent.com/263325/177266231-61a94b05-ee9f-46b4-9080-10c51dc67fbf.png">

**Reply by raydeleu** [reply]
I tested this solution. It works, but still I have the impression that the solution with `cubicBezierCurve` works better in this case. For the model I assumed that the user will not adapt the curve, only the dimensions of part that hold the earplug. If the user wants to adapt the shape of the curve, your solution provides more flexibility. Using normal circular curves could also do the job. We could also consider whether the `Sketcher` or `draw()` should offer the option to fillet a curve.

**Comment by raydeleu** [comment]
<img width="372" alt="morepoints" src="https://user-images.githubusercontent.com/38007983/177217344-6230452e-0d42-4a5b-b8aa-2611299ca2ce.png">

I have to correct myself: adding more points is certainly not an easy solution to create a better curve with drawPointsInterpolation. If you add too many points and these are not accurate, the curve "goes crazy".

**Reply by sgenoud** [reply]
There is also the solution of changing the interpolation options. For instance you can raise the degMax of the spline, which might break SVG export.

**Comment by raydeleu** [comment]
The `drawPointsInterpolation` works very nicely to create an airfoil shape. I created a random airfoil data file and manually converted the data file into an array. I then applied the `drawPointsInterpolation` to this complete array. The coordinate file contains the coordinates of points along the contour, starting and ending at the sharp trailing edge. 

<img width="604" alt="airfoil-replicad" src="https://user-images.githubusercontent.com/38007983/177646351-67635dff-9010-4246-93c2-b2ba56508a99.png">

The code is: 

```` javascript
// Replicad code to draw an airfoil shape
// Purpose: demonstrate how to draw an airfoil
// airfoil design from https://www.flusur.com/airfoil-design/

function main(
 { drawPointsInterpolation })   // functions used within the main program  						

{

// the following array of points was created from a 
// standard airfoil coordinate data file
// the airfoil coordinate file is normalised to a standard chord of 1

let airfoilPoints=
[[1.0000000   ,0.0000000],
[0.9985383   ,0.0003261],
[0.9943357   ,0.0012896],
[0.9874835   ,0.0028991],
[0.9780353   ,0.0051591],
[0.9660390   ,0.0080639],
[0.9515443   ,0.0115947],
[0.9346057   ,0.0157195],
[0.9152843   ,0.0203926],
[0.8936495   ,0.0255547],
[0.8697794   ,0.0311333],
[0.8437625   ,0.0370433],
[0.8156986   ,0.0431884],
[0.7857000   ,0.0494617],
[0.7538929   ,0.0557483],
[0.7204180   ,0.0619265],
[0.6854321   ,0.0678706],
[0.6491079   ,0.0734536],
[0.6116358   ,0.0785498],
[0.5732228   ,0.0830380],
[0.5340930   ,0.0868053],
[0.4944863   ,0.0897496],
[0.4546574   ,0.0917838],
[0.4148737   ,0.0928383],
[0.3754127   ,0.0928639],
[0.3365591   ,0.0918344],
[0.2986011   ,0.0897478],
[0.2618267   ,0.0866279],
[0.2265198   ,0.0825240],
[0.1929555   ,0.0775104],
[0.1613969   ,0.0716848],
[0.1320908   ,0.0651660],
[0.1052651   ,0.0580905],
[0.0811259   ,0.0506086],
[0.0598565   ,0.0428803],
[0.0416163   ,0.0350696],
[0.0265420   ,0.0273396],
[0.0147495   ,0.0198457],
[0.0063383   ,0.0127275],
[0.0013965   ,0.0060949],
[0.0000000   ,0.0000000],
[0.0025589   ,-0.0060606],
[0.0092026   ,-0.0115698],
[0.0196689   ,-0.0165551],
[0.0336781   ,-0.0209976],
[0.0509521   ,-0.0248665],
[0.0712157   ,-0.0281356],
[0.0941973   ,-0.0307883],
[0.1196300   ,-0.0328200],
[0.1472530   ,-0.0342383],
[0.1768128   ,-0.0350618],
[0.2080639   ,-0.0353198],
[0.2407701   ,-0.0350502],
[0.2747044   ,-0.0342981],
[0.3096492   ,-0.0331148],
[0.3453962   ,-0.0315557],
[0.3817458   ,-0.0296795],
[0.4185066   ,-0.0275463],
[0.4554945   ,-0.0252170],
[0.4925323   ,-0.0227516],
[0.5294482   ,-0.0202088],
[0.5660754   ,-0.0176447],
[0.6022513   ,-0.0151121],
[0.6378161   ,-0.0126595],
[0.6726125   ,-0.0103308],
[0.7064852   ,-0.0081646],
[0.7392796   ,-0.0061931],
[0.7708424   ,-0.0044425],
[0.8010203   ,-0.0029317],
[0.8296603   ,-0.0016722],
[0.8566098   ,-0.0006682],
[0.8817164  ,0.0000844],
[0.9048280  ,0.0005973],
[0.9257939  ,0.0008906],
[0.9444642  ,0.0009917],
[0.9606910  ,0.0009361],
[0.9743280  ,0.0007659],
[0.9852307  ,0.0005300],
[0.9932535  ,0.0002828],
[0.9982433  ,0.0000842],
[1.0000000  ,0.0000000]]

// scale chord length
let chord = 100 
let span  = 100
let airfoilPointsLarge = airfoilPoints.map(function([x,y]){return [x*chord,y*chord]}) 
let airfoil = drawPointsInterpolation(airfoilPointsLarge).sketchOnPlane("XZ");
let wing = airfoil.extrude(span)

return  [wing]

}
````
